### PR TITLE
Added extended entitlements

### PR DESCRIPTION
**Pull request by BaranekD**

### CODE REVIEW

File: lib/EntitlementUtils.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+ */
+class EntitlementUtils
+{
+    public static function getForwardedEduPersonEntitlement(&$request, $adapter, $forwardedEduPersonEntitlement)
+    {
+        $result = [];
+
+        if (!isset($request['perun']['user'])) {
+            Logger::debug(
+                'perun:PerunEntitlement: Object Perun User is not specified.' .
```
Comment (vyskocilpavel) [inline lib/EntitlementUtils.php:24]: Should be ClassName:methodName

File: lib/EntitlementUtils.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        $user = $request['perun']['user'];
+
+        try {
+            $forwardedEduPersonEntitlementMap = $adapter->getUserAttributesValues(
+                $user,
+                [$forwardedEduPersonEntitlement]
+            );
+        } catch (Exception $exception) {
+            Logger::error(
+                'perun:PerunEntitlement: Exception ' . $exception->getMessage() .
```
Comment (vyskocilpavel) [inline lib/EntitlementUtils.php:39]: Should be ClassName:methodName

File: lib/EntitlementUtils.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        $facilityCapabilities = [];
+        $capabilitiesResult = [];
+
+        $spEntityId = self::getSpEntityId($request);
+        try {
+            $resourceCapabilities = $adapter->getResourceCapabilities($spEntityId, $request['perun']['groups']);
+            $facilityCapabilities = $adapter->getFacilityCapabilities($spEntityId);
+        } catch (Exception $exception) {
+            Logger::error(
+                'perun:PerunEntitlement: Exception ' . $exception->getMessage() .
```
Comment (vyskocilpavel) [inline lib/EntitlementUtils.php:63]: Should be ClassName:methodName

File: lib/EntitlementUtils.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        return $name;
+    }
+
+    private static function getSpEntityId(&$request)
+    {
+        if (isset($request['SPMetadata']['entityid'])) {
+            return $request['SPMetadata']['entityid'];
+        } else {
+            throw new Exception('perun:PerunEntitlement: Cannot find entityID of remote SP. ' .
```
Comment (vyskocilpavel) [inline lib/EntitlementUtils.php:127]: Should be ClassName:methodName

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+            $capabilities = EntitlementUtils::getCapabilities(
+                $request,
+                $this->adapter,
+                $this->entitlementPrefix,
+                $this->entitlementAuthority
+            );
+        } else {
+            Logger::debug(
+                'perun:PerunEntitlement: There are no user groups assign to facility.' .
```
Comment (vyskocilpavel) [inline lib/Auth/Process/PerunEntitlementExtended.php:90]: Should be perun:PerunEntitlementExtended

**Review by vyskocilpavel — CHANGES_REQUESTED** [review]

**Review by vyskocilpavel — APPROVED** [review]

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+            $entitlementWithAttributes = $this->groupEntitlementWithAttributesWrapper($group->getUuid(), $groupName);
+            array_push($eduPersonEntitlementExtended, $entitlementWithAttributes);
+        }
+
+        natsort($eduPersonEntitlementExtended);
+        return $eduPersonEntitlementExtended;
+    }
+
+    private function groupNameWrapper($groupName)
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:132]: I don't like the "groupName" wording there. Clearly, we are putting the UUID of the group as a parameter. As an external, I would probably get confused why this is happenning.

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        return $this->entitlementPrefix . 'group:' . EntitlementUtils::encodeName($groupName) .
+            '#' . $this->entitlementAuthority;
+    }
+
+    private function groupEntitlementWrapper($groupName)
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:138]: Same as above.

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            array_push($eduPersonEntitlementExtended, $entitlementWithAttributes);
+        }
+
+        natsort($eduPersonEntitlementExtended);
+        return $eduPersonEntitlementExtended;
+    }
+
+    private function groupNameWrapper($groupName)
+    {
+        return $this->entitlementPrefix . 'group:' . EntitlementUtils::encodeName($groupName) .
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:134]: perhaps extracting the "group" as a constant would be a good idea.

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            '#' . $this->entitlementAuthority;
+    }
+
+    private function groupEntitlementWrapper($groupName)
+    {
+        return $this->entitlementPrefix . 'group:' . $groupName . '#' . $this->entitlementAuthority;
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:140]: Why isn't the groupName encoded? In the method above and below, we are encoding it, and here not.

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    private function groupEntitlementWithAttributesWrapper($uuid, $groupName)
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:143]: I think these wrapper methods should be extracted into the "utils" class as well as in the case of the capabilities wrapper.

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    private function getEduPersonEntitlementExtended(&$request)
+    {
+        $eduPersonEntitlementExtended = [];
+
+        $groups = $request['perun']['groups'];
+        foreach ($groups as $group) {
+            array_push($eduPersonEntitlementExtended, $this->groupNameWrapper($group->getUuid()));
+
+            $groupName = $group->getUniqueName();
+            $groupName = preg_replace('/^(\w*)\:members$/', '$1', $groupName);
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:119]: I would move these two lines either into the "wrapper" or closer to line 124 (as it uses it as a passed parameter).

File: lib/Auth/Process/PerunEntitlementExtended.php [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        $request['Attributes'][$this->outputAttrName] = array_unique(array_merge(
+            $eduPersonEntitlementExtended,
+            $forwardedEduPersonEntitlement,
+            $capabilities
+        ));
+    }
+
+    private function getEduPersonEntitlementExtended(&$request)
```
Comment (dBucik) [inline lib/Auth/Process/PerunEntitlementExtended.php:110]: I see three output values for one group:
```
group:NAME
group:UUID
groupAttributes:UUID?...
```
Wasn't it agreed to leave out the NAME version?

**Review by dBucik — CHANGES_REQUESTED** [review]

**Comment by BaranekD** [comment]
Requested changes done. I've also squashed commits into one.

**Review by dBucik — APPROVED** [review]

**Pull request merged**